contracts/Nonon/NononFriendCard.sol

```diff
@@ -0,0 +1,336 @@
+            }
+        }
+    }
+
+    // prevent transfer (except mint and burn)
+    function _beforeTokenTransfers(address from, address to, uint256, uint256) internal pure override {
+        if (from != address(0) && to != address(0)) {
+            revert OnlyForYou();
+        }
+    }
+
+    // add ID for associated sequential tokens to appropriate lists
+    function registerTokenMovement(address from, address to, uint256 collectionTokenStartId, uint256 quantity)
+        external
+        override
+        onlyCollection
+    {
+        if (from != address(0)) {
+            if (to != from) {
+                sentBitmap[from].setBatch(collectionTokenStartId, quantity);
+            }
+        }
+
+        if (to != address(0)) {
+            receivedBitmap[to].setBatch(collectionTokenStartId, quantity);
+        }
+
+        emit BatchMetadataUpdate(1, type(uint256).max);
+    }
+
+    // total points accumulated by a holder
+    function points(uint256 tokenId) public view returns (uint256) {
+        address owner = ownerOf(tokenId);
+        uint256 max = IERC721A(collectionAddress).totalSupply() + 1;
+
+        return receivedBitmap[owner].popCount(1, max) + sentBitmap[owner].popCount(1, max);
+    }
+
+    // convenience function to get point information in a token range
+    // note that this is expensive and most likely will require multiple calls to cover large ranges.
+    function tokenPointsInRange(uint256 startId, uint256 endId) public view returns (TokenPoints[] memory) {
+        if (endId < startId) revert InvalidParams();
+
+        TokenPoints[] memory tokenPoints = new TokenPoints[]((endId - startId) + 1);
+        uint256 max = IERC721A(collectionAddress).totalSupply() + 1;
+
+        uint256 pointsIndex;
+        for (uint256 i = startId; i <= endId;) {
+            if (_exists(i)) {
+                address owner = ownerOf(i);
+                uint256 totalPoints = receivedBitmap[owner].popCount(1, max) + sentBitmap[owner].popCount(1, max);
+
+                tokenPoints[pointsIndex] = TokenPoints({id: i, owner: owner, points: totalPoints});
+                ++pointsIndex;
+            }
+            ++i;
+        }
+
+        return tokenPoints;
+    }
+
+    // check if given address is a holder of the token
+    function hasToken(address receiver) public view override returns (bool) {
+        return balanceOf(receiver) > 0;
+    }
+
+    // check if given address has ever received tokenId
+    function hasReceivedToken(address owner, uint256 tokenId) external view returns (bool) {
+        return receivedBitmap[owner].get(tokenId);
+    }
+
+    // check if given address has ever sent tokenId
+    function hasSentToken(address owner, uint256 tokenId) external view returns (bool) {
+        return sentBitmap[owner].get(tokenId);
+    }
+
+    function tokenStatusMap(address owner, bool sent) external view returns (uint256[] memory received) {
+        // TODO: reference max supply instead of hardcoding
```

Comment:
Reminder for this todo

Reply:
went with a constant in this case, added it and removed comment in https://github.com/guminc/contracts/pull/73/commits/67285f429a95c1766f6f0e11e1d2c44eb1eeb991